django/test/utils.py

```diff
@@ -381,6 +382,54 @@ def get_runner(settings, test_runner_class=None):
     return getattr(test_module, test_path[-1])
 
 
+def run_pairing(paired_test, test_labels, subprocess_args):
+    for i, label in enumerate(test_labels):
+        print(
+            "***** %d of %d: Check test pairing with %s"
+            % (i + 1, len(test_labels), label)
+        )
+        failures = subprocess.call(subprocess_args + [label, paired_test])
+        if failures:
+            print("***** Found problem pair with %s" % label)
+            return
+
+    print("***** No problem pair found")
+
+
+def run_bisection(bisection_label, test_labels, subprocess_args):
+    iteration = 1
+    while len(test_labels) > 1:
+        midpoint = len(test_labels) // 2
+        test_labels_a = test_labels[:midpoint] + [bisection_label]
+        test_labels_b = test_labels[midpoint:] + [bisection_label]
+        print("***** Pass %da: Running the first half of the test suite" % iteration)
+        print("***** Test labels: %s" % " ".join(test_labels_a))
+        failures_a = subprocess.run(subprocess_args + test_labels_a)
+
+        print("***** Pass %db: Running the second half of the test suite" % iteration)
+        print("***** Test labels: %s" % " ".join(test_labels_b))
+        print("")
+        failures_b = subprocess.run(subprocess_args + test_labels_b)
+
+        if failures_a.returncode and not failures_b.returncode:
+            print("***** Problem found in first half. Bisecting again...")
+            iteration += 1
+            test_labels = test_labels_a[:-1]
+        elif failures_b.returncode and not failures_a.returncode:
+            print("***** Problem found in second half. Bisecting again...")
+            iteration += 1
+            test_labels = test_labels_b[:-1]
+        elif failures_a.returncode and failures_b.returncode:
+            print("***** Multiple sources of failure found")
+            break
+        else:
+            print("***** No source of failure found... try pair execution (--pair)")
+            break
+
+    if len(test_labels) == 1:
+        print("***** Source of error: %s" % test_labels[0])
```

Comment:
You copied a lot of code from `runtests.py`. As far as I'm aware, this patch should removed the matching code from `runtests.py` and use the functionality that we introduced in `DiscoverRunner`.